src/mapping/mapping_helpers.py

```diff
@@ -2,47 +2,116 @@
+        mapper_name (str): The name of the mapper being validated.
+        col (str): The name of the column to check.
+        expected_value_type (type): The expected type of the column values.
+        expected_value_length (int): The expected length of the column values.
+        check_capitalisation (bool, optional): Whether to check if the column
+        values are in uppercase. Defaults to False.
+
+    Raises:
+        ValueError: If any of the validation checks fail.
+
     """
     try:
-        # Check DataFrame shape
-        if df.shape[1] != 2:
-            raise ValueError("Dataframe file must have exactly two columns")
+        if expected_value_type is not None:
+            # Check datatype correct
+            if not df[col][df[col].notnull()].apply(type).eq(expected_value_type).all():
+                raise ValueError(
+                    f"Column {col} is not of the expected type {expected_value_type}."
+                )
+        if expected_value_length is not None:
+            # Convert to strings, and check if value lengths are correct
+            if (
+                not df[col][df[col].notnull()]
+                .astype(str)
+                .str.len()
+                .eq(expected_value_length)
+                .all()
+            ):
+                raise ValueError(
+                    f"Column {col} does not have the expected length of"
+                    f" {expected_value_length}."
+                )
+        if check_capitalisation:
+            # Check if all non-null values in the column are in uppercase
+            if not df[col][df[col].notnull()].str.isupper().all():
+                raise ValueError(f"Column {col} is not in uppercase.")
+    except ValueError as ve:
+        raise ValueError(f"{mapper_name} mapper validation failed: " + str(ve))
 
-        # Check column headers
-        if list(df.columns) != ["ruref", "ultfoc"]:
-            raise ValueError("Column headers should be 'ruref' and 'ultfoc'")
 
-        # Check 'ultfoc' values are either 2 characters or 'nan'
-        def check_ultfoc(value):
-            if pd.isna(value):
-                return True
-            else:
-                return isinstance(value, str) and (len(value) == 2)
+def check_mapping_unique(
+    df: pd.DataFrame,
+    col_to_check: str,
+) -> None:
+    """
+    Checks if a column contains unique values.
 
-        df["contents_check"] = df.apply(lambda row: check_ultfoc(row["ultfoc"]), axis=1)
+    Parameters:
+    df (pd.DataFrame): The DataFrame to check.
+    col_to_check (str): The name of the column to check.
 
-        # check any unexpected contents
-        if not df["contents_check"].all():
-            raise ValueError("Unexpected format within 'ultfoc' column contents")
+    Raises:
+    ValueError: If the column does not contain unique values.
+    """
+    if not df[col_to_check].is_unique:
+        raise ValueError(f"Column {col_to_check} is not unique.")
```

Comment:
Have you tested that this error is raised? if not, leave for now.

Reply:
There's a unit test to check, and just to double check, I've just run it paying with the test data and the error is raised